src/common/utils.h

```diff
@@ -43,9 +43,157 @@
 #include <algorithm>
 #include <functional>
 
+#include "../operator/mxnet_op.h"
+
 namespace mxnet {
 namespace common {
 
+
+/*!
+ * \brief IndPtr should be in non-decreasing order, start with 0
+ *           and end with value greater or equal than size of indices.
+ */
+struct csr_indptr_check {
+  template<typename DType, typename IType>
+  MSHADOW_XINLINE static void Map(int i, DType* out, const IType* in,
+                                  const nnvm::dim_t end, const nnvm::dim_t idx_size) {
+    if ((in[i+1] < in[i]) || (i == 0 && in[i] != static_cast<IType>(0)) ||
+        (i == end && in[i] < static_cast<IType>(idx_size)))
+      *out = kCSRIndPtrErr;
+  }
+};
+
+/*!
+ *  \brief Indices should be less than the number of columns.
+ */
+struct csr_idx_check {
+  template<typename DType, typename IType>
+  MSHADOW_XINLINE static void Map(int i, DType* out,
+                                  const IType* in, const nnvm::dim_t ncols) {
+    if (in[i] >= static_cast<IType>(ncols)) *out = kCSRIdxErr;
```

Comment:
We should also check in[i] >=0, and csr_indices are expected to be in sorted order per row:
https://mxnet.incubator.apache.org/versions/master/api/python/ndarray/sparse.html#mxnet.ndarray.sparse.CSRNDArray 
We should also add test for this. 
e.g. this is invalid CSR in MXNet:
shape = (1, 3)
data=[1,2,3]
indptr = [0,3]
indices = [3,2,1] --> not in ascending order

Reply:
I found yesterday that scipy will not keep the order of the indices[indptr[i]:indptr[i+1]] if we slice the sparse matrix using M[row_indices,:][:, col_indices].